preflight_check/task.py

```diff
@@ -0,0 +1,484 @@
+        1. There is one latest fence action successfully done
+        2. No fence action during fence timeout, thread_stop_event triggered by main thread
+        """
+        target_node = None
+        from_node = None
+        fence_timestamp = None
+
+        # Try to find out which node fire the fence action
+        while not self.thread_stop_event.is_set():
+            rc, out, _ = crmshutils.get_stdout_stderr("crm_mon -1|grep -A1 \"Fencing Actions:\"")
+            if rc == 0 and out:
+                match = re.search(r"of (.*) pending: .*origin=(.*)$", out)
+                if match:
+                    target_node, from_node = match.groups()
+                    self.info("Node \"{}\" will be fenced by \"{}\"!".format(target_node, from_node))
+                    self.fence_start_event.set()
+                    break
+            time.sleep(1)
+
+        # Try to find out proof that fence happened
+        while not self.thread_stop_event.is_set():
+            rc, out, _ = crmshutils.get_stdout_stderr(config.FENCE_HISTORY.format(node=target_node))
+            if rc == 0 and out:
+                match = re.search(r"Node {} last fenced at: (.*)".format(target_node), out)
+                if match:
+                    fence_timestamp = match.group(1)
+                    task_timestamp_dt = utils.str_to_datetime(self.timestamp, '%Y/%m/%d %H:%M:%S')
+                    fence_timestamp_dt = utils.str_to_datetime(fence_timestamp, '%a %b %d %H:%M:%S %Y')
+                    # If the fence action timestamp larger than this task's timestamp
+                    # That is the proof
+                    if task_timestamp_dt < fence_timestamp_dt:
+                        self.info("Node \"{}\" was successfully fenced by \"{}\"".format(target_node, from_node))
+                        # Tell main thread fence happened
+                        self.fence_finish_event.set()
+                        break
+            time.sleep(1)
+
+
+class TaskFence(Task):
+    """
+    Class to fence node
+    """
+    def  __init__(self, target):
+        """
+        Init function
+        """
+        self.target_node = target
+        description = "Fence node {}".format(self.target_node)
+        super(self.__class__, self).__init__(description, flush=True)
+
+    def header(self):
+        """
+        Header content for this task
+        """
+        h = '''==============================================
+Testcase:          {}
+Fence action:      {}
+Fence timeout:     {}
+'''.format(self.description, self.fence_action, self.fence_timeout)
+        return h
+
+    def to_json(self):
+        """
+        Dump join result
+        """
+        self.build_base_result()
+        self.result['Fence action'] = self.fence_action
+        self.result['Fence timeout'] = self.fence_timeout
+        from . import main
+        main.ctx.task_list = self.prev_task_list + [self.result]
+        utils.json_dumps()
+
+    def to_report(self):
```

Comment:
Put these empty function `to_report` from each `TaskXXX` object to the `Task`.

Reply:
Changed.
Thanks!